_posts/2017-04-02-generating-sitemaps-with-apache-spark.markdown

```diff
@@ -0,0 +1,228 @@
+
+<!-- more -->
+
+### What's in a Sitemap?
+
+If you're not familiar with sitemaps, they are an easy way for us to inform search engines about pages on artsy.net available for crawling. A Sitemap is an XML file that lists URLs along with some additional metadata. All Artsy sitemaps are listed in our [robots.txt](https://www.artsy.net/robots.txt).
+
+```
+User-agent: *
+Sitemap: https://www.artsy.net/sitemap-artists.xml
+Sitemap: https://www.artsy.net/sitemap-shows.xml
+...
+```
+
+These are actually sitemap indexes, such as [the shows sitemap index](https://www.artsy.net/sitemap-shows.xml). Each index contains links to the actual sitemaps.
+
+```xml
+<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+ <sitemap>
+  <loc>https://www.artsy.net/sitemap-shows-2016.xml</loc>
+ </sitemap>
+ ...
+</sitemapindex>
+```
+
+Each sitemap contains links to web pages on [www.artsy.net](https://www.artsy.net), such as the [sitemap containing links to shows in 2016](https://www.artsy.net/sitemap-shows-2016.xml).
+
+```xml
+<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+ <url>
+  <loc>https://www.artsy.net/show/guy-hepner-kate-moss-exhibiton</loc>
+ </url>
+ ...
+</urlset>
+```
+
+### The Monolith's Rake Task
+
+When Artsy was a Ruby on Rails monolith we generated and wrote static sitemap files to S3 in a Rake task. We iterated over all our data nightly, created a set of URLs and wrote an XML file. We got fairly clever with a `Sitemappable` module that was included in anything that yielded a sitemap.
+
+```ruby
+module Sitemappable
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def sitemappable_priority
+      0.5
+    end
+  end
+end
+```
+
+We also used HAML to generate sitemaps.
+
+```haml
+!!! XML
+%urlset{ xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9" }
+  - documents.each do |document|
+      - type.sitemappable_urls(document, self).each do |url|
+        %url
+          %loc=h url
+```
+
+Of course, there were many more implementation details, but adding a sitemap was easy as we made Mongoid models use `Sitemappable`.
+
+Generating sitemaps took a couple of hours and ran nightly.
+
+### The Dynamic Front-End
+
+By 2015 our front-end was fully decoupled from the monolith that now served as the core API service. The sitemap code still lived there and needed to know about an increasing number of URLs, which seemed totally backwards. We wanted an easier way to maintain and update sitemaps and concluded that by generating sitemaps on the fly we would also see fewer 404 errors during crawling as a side effect. Furthermore, sitemaps increasingly needed information from other places, as we tore the monolith apart into micro-services.
```

Comment:
`...also see fewer 404 errors during crawling...` might be unclear. Maybe explain that generating real-time allows them to reflect our live data more accurately?